src/main/java/io/cdap/plugin/http/source/common/pagination/page/BasePage.java

```diff
@@ -42,14 +56,78 @@ public BasePage(BaseHttpSourceConfig config, HttpErrorHandler httpErrorHandler,
 
   @Nullable
   public abstract String getPrimitiveByPath(String path);
+  public abstract boolean hasNext();
+  protected abstract StructuredRecord next() throws TransformationException;
 
   public HttpErrorHandlingStrategy getErrorHandlingStrategy() {
     return httpErrorHandler
       .getErrorHandlingStrategy(httpResponse.getStatusCode())
       .getAfterRetryStrategy();
   }
 
-  public HttpResponse getHttpResponse() {
-    return httpResponse;
+  /**
+   * Run functions for every record on the page if appropriate. Or for whole page if has erroneous http
+   * status code.
+   *
+   * @param onOutput is run for non-erroneous records on the page
+   * @param onSendToError is run for erroneous records with "send to error" strategy. Or once for whole page with
+   *                      erroneous http status code and strategy "send to error".
+   */
+  public void consume(Consumer<StructuredRecord> onOutput, Consumer<InvalidEntry<StructuredRecord>> onSendToError)
+    throws TransformationException {
+
+    if (handleHttpError(onSendToError)) {
+      return;
+    }
+
+    while (hasNext()) {
+      try {
+        onOutput.accept(next());
+      } catch (TransformationException ex) {
+        String transformedString = ex.getTransformedString();
+        Throwable cause = ex.getCause();
+
+        switch (config.getErrorHandling()) {
+          case SKIP:
+            LOG.warn("Cannot convert row '{}' to a record", transformedString, cause);
+            break;
+          case SEND:
+            onSendToError.accept(buildError(0, transformedString,
+                                         cause.getClass().getName() + ": " + cause.getMessage()));
+            break;
+          case STOP:
+            throw ex;
+          default:
+            throw new UnexpectedFormatException(
+              String.format("Unknown error handling strategy '%s'", config.getErrorHandling()));
+        }
+      }
+    }
+  }
+
+  private boolean handleHttpError(Consumer<InvalidEntry<StructuredRecord>> onSendToError) {
```

Comment:
Instead of having this logic here, its cleaner if there is an ErrorPage class instead of having this logic mixed in the base class. ErrorPage would emit an InvalidEntry in it's consume() method. The page iterator should create the page if the request failed and it is configured to send to error.

This separates the logic more cleanly and prevents somebody from overwriting the consume() method and forgetting to handle errors.

Reply:
Since we need Page to be iterator (read the comment below, where I explained it). I don't think it's really easy (or makes sense) to implement this.